physics.quantum: Implemented _represent_ZGate in Grover's algorithm

Temigo opened the pull request:
Implemented function _represent_ZGate in grover.py : it computes
the representation of the OracleGate in the computational basis.

There might be a cleverer way of doing it, but currently it just
looks at the oracle answer for every qubit of the computational
basis.

Temigo committed:
physics.quantum: Implemented _represent_ZGate in Grover's algorithm

Implemented function _represent_ZGate in grover.py : it computes
the representation of the OracleGate in the computational basis.
Tests will soon be added.

moorepants commented:
Hi Temigo. I don't think we have many people actively developing the physics quantum stuff at the moment. I'd recommend using git blame on these files and then @mentioning the main developers. You are more likely to get faster feedback that way.

Temigo commented:
Thanks for the advice ! I think that @ellisonbg and @rwong are the more concerned by this pull request.

rwong commented on sympy/physics/quantum/grover.py:
```diff
@@ -168,9 +169,17 @@ def _apply_operator_Qubit(self, qubits, **options):
+        """
+        nbasis = 2**self.nqubits  # compute it only once
+        matrixOracle = eye(nbasis)
+        normalizing_factor = 1/sqrt(nbasis)
```
The matrix rep actually doesn't have the normalization factor (repr(oracle) = I - 2|v><v|, where |v> is the target computational basis state).

Temigo committed:
physics.quantum: Corrected _represent_ZGate of Grover

Error in function _represent_ZGate fixed in Grover's algorithm
(normalisation factor).

Temigo commented:
Oops. My mistake, sorry for that. Should be corrected by now.

rwong commented:
I'm fine with this implementation for now but I think it's a good idea to wait for feedback from Brian (@ellisonbg) as well.

certik commented:
I think @ellisonbg is busy. I am merging this.

ellisonbg commented:
+1

On Wed, Dec 16, 2015 at 1:33 PM, Ondřej Čertík <notifications@github.com>
wrote:

> Merged #9137 <https://github.com/sympy/sympy/pull/9137>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9137#event-493982431>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
